packages/@lwc/ssr-compiler/src/shared.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+export type Expression = string;
+
+export type Instruction =
+    | IEmitTagName
+    | IEmitStaticString
+    | IEmitExpression
+    | IStartConditional
+    | IEndConditional
+    | IInvokeConnectedCallback
+    | IRenderChild
+    | IHoistImport
+    | IHoistInstantiation;
+
+export interface IEmitTagName {
+    kind: 'emitTagName';
+}
+
+export interface IEmitStaticString {
+    kind: 'emitStaticString';
+}
+
+export interface IEmitExpression {
+    kind: 'emitExpression';
+    expression: Expression;
+}
+
+export interface IStartConditional {
+    kind: 'startConditional';
+}
+
+export interface IEndConditional {
+    kind: 'endConditional';
+}
+
+export interface IInvokeConnectedCallback {
+    kind: 'invokeConnectedCallback';
+}
+
+export interface IRenderChild {
+    kind: 'renderChild';
+    dynamic: Expression | null;
+}
+
+export interface IHoistImport {
+    kind: 'hoistImport';
+}
+
+export interface IHoistInstantiation {
+    kind: 'hoistInstantiation';
+}
+
+export type GenerateMarkupFn = (
+    tagName: string,
+    props: Record<string, any> | null,
+    attrs: Record<string, any> | null,
+    slotted: Record<number | string, AsyncGenerator<string>> | null
+) => AsyncGenerator<string>;
+
+export type TemplateFn = (Ctor: any, instance: any | null) => AsyncGenerator<string>;
```

Comment:
We don't accomplish much with `unknown` other than forcing consumers to cast the `unknown` to `any` in their codebase. When dealing with generated code, there are boundaries in our own codebase beyond which referential integrity goes away. This is largely because the interaction occurs post-build, where JavaScript is interacting with JavaScript and the types are not known ahead of time.

In my mind, this is actually a good case for `any`, whereas usually I'd prefer unknown.